optimum/onnxruntime/modeling_decoder.py

```diff
@@ -695,3 +696,54 @@ def _reorder_cache(past: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor) ->
     def can_generate(self):
         """Returns True to validate the check that the model using `GenerationMixin.generate()` can indeed generate."""
         return True
+
+    @classmethod
+    def _from_pretrained(
+        cls,
+        model_id: Union[str, Path],
+        config: "PretrainedConfig",
+        **kwargs,
+    ):
+        if config.model_type == "bloom":
+            return super()._from_pretrained(model_id, config, init_cls=ORTBloomForCausalLM, **kwargs)
+        else:
+            return super()._from_pretrained(model_id, config, init_cls=ORTModelForCausalLM, **kwargs)
+
+
+class ORTBloomForCausalLM(ORTModelForCausalLM):
+    # Adapted from transformers.models.bloom.modeling_bloom.BloomForCausalLM._reorder_cache
+    def prepare_inputs_for_generation(self, input_ids, past_key_values=None, **kwargs):
+        attention_mask = kwargs.get("attention_mask", None)
+        use_cache = kwargs.get("use_cache", None)
+
+        # only last token for input_ids if past is not None
+        if past_key_values:
+            # the cache may be in the stardard format (e.g. in contrastive search), convert to bloom's format if needed
+            if past_key_values[0][0].shape[0] == input_ids.shape[0]:
+                past_key_values = BloomForCausalLM._convert_to_bloom_cache(past_key_values)
```

Comment:
Not a super satisfying approach but here is what we've done for OpenVINO : https://github.com/huggingface/optimum-intel/blob/v1.9.3/optimum/intel/openvino/modeling_decoder.py#L322 in order to rely as little as possible on `transformers` (which will be a dependency until we've our own `.generate()` so doesn't change much for now)

Reply:
Thanks! I see the bug has been fixed already in optimum-intel. Feel free to ping me / open an issue in optimum in such case, as it's likely the bug is there as well. Trying to do as well!

Reply:
> Thanks! I see the bug has been fixed already in optimum-intel. Feel free to ping me / open an issue in optimum in such case, as it's likely the bug is there as well. Trying to do as well!

This was integrated in https://github.com/huggingface/optimum-intel/pull/274 when enabling past key values re-use, so never saw that as a bug per say that could also happen in optimum, otherwise always do !> I'm personally not super fan of if/else, I think in this case subclassing makes sense (though here it is still ugly). But I'll remove the import indeed!

Reply:
> Thanks! I see the bug has been fixed already in optimum-intel. Feel free to ping me / open an issue in optimum in such case, as it's likely the bug is there as well. Trying to do as well!

This was integrated in https://github.com/huggingface/optimum-intel/pull/274 when enabling past key values re-use for openvino models, so never saw that as a bug to be fixed in optimum as well, otherwise always do!